File: solutions/0/0/5/2/p0052_s0.c
/* Solution to Euler Problem 52: Permuted Multiples. */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

static int cmp_char(const void *a, const void *b) {
    return (*(const char *)a) - (*(const char *)b);
}

static void sorted_digits(long long n, char *buf) {
    /* Write sorted decimal digits of n into buf (null-terminated). */
    char tmp[32];
    int len = 0;
    if (n == 0) {
        buf[0] = '0';
        buf[1] = '\0';
        return;
    }
    while (n > 0) {
        tmp[len++] = (char)('0' + (n % 10));
        n /= 10;
    }
    qsort(tmp, (size_t)len, 1, cmp_char);
    memcpy(buf, tmp, (size_t)len);
    buf[len] = '\0';
}

long long solve(int argc, char *argv[]) {
    int multiples = 6;
    if (argc >= 2) {
        multiples = atoi(argv[1]);
    }
    if (multiples < 2 || multiples > 6) {
        fprintf(stderr, "multiples must be between 2 and 6 inclusive.\n");
        return -1;
    }

    char fingerprint[32];
    char candidate[32];

    for (long long i = 1; ; i++) {
        /* Compute sorted-digit fingerprint of i */
        sorted_digits(i, fingerprint);

        int all_match = 1;
        for (int m = 2; m <= multiples; m++) {
            sorted_digits(i * (long long)m, candidate);
            if (strcmp(fingerprint, candidate) != 0) {
                all_match = 0;
                break;
            }
        }

        if (all_match) {
            return i;
        }
    }
}

/* Usage: ./file <kwarg>... [--runs=1] [--show]
 * Output: "<runs> <avg_seconds> <result>" */
int main(int argc, char *argv[]) {
    int runs = 1;

    char **solve_argv = malloc((size_t)argc * sizeof(char *));
    if (!solve_argv) {
        fprintf(stderr, "runner: out of memory\n");
        return 1;
    }
    int solve_argc = 0;
    solve_argv[solve_argc++] = argv[0];

    for (int i = 1; i < argc; i++) {
        if (argv[i][0] == '\0') continue;
        if (strncmp(argv[i], "--runs=", 7) == 0) {
            int r = atoi(argv[i] + 7);
            if (r >= 1) runs = r;
            continue;
        }
        if (strcmp(argv[i], "--show") == 0) continue;
        solve_argv[solve_argc++] = argv[i];
    }

    long long result = 0;
    double total = 0.0;
    int rc = 0;
    int has_result = 0;

    for (int r = 0; r < runs; r++) {
        struct timespec t0, t1;
        clock_gettime(CLOCK_MONOTONIC, &t0);
        long long cur = solve(solve_argc, solve_argv);
        clock_gettime(CLOCK_MONOTONIC, &t1);
        total += (double)(t1.tv_sec - t0.tv_sec)
               + (double)(t1.tv_nsec - t0.tv_nsec) * 1e-9;
        if (has_result && cur != result) {
            fprintf(stderr, "Expected consistent result, got %lld previous result=%lld\n",
                    cur, result);
            rc = 1;
        }
        result = cur;
        has_result = 1;
    }

    free(solve_argv);
    printf("%d %.17g %lld\n", runs, total / (double)runs, result);
    return rc;
}

File: solutions/0/0/5/2/p0052_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 52: Permuted Multiples [Level 0]. """
from __future__ import annotations

import sys
from sys import argv, stderr
from time import perf_counter
from typing import Any


def solve(*, multiples: int) -> int:
    if not (isinstance(multiples, int) and 1 < multiples < 7):
        raise ValueError("multiples must be an integer between 2 and 6, both inclusive.")
    multiples_range = tuple(range(1, multiples + 1))
    for i in range(1, sys.maxsize // multiples):
        if len({"".join(sorted(str(i * multiple))) for multiple in multiples_range}) == 1:
            return i
    else:
        raise ValueError("No solution found")


def main(**kwargs: Any) -> int:
    """
    Usage: ./file.py <kwarg>... [--runs=1] [--show]
    Output: "<runs> <avg_seconds> <result>"
    """
    try:
        runs_arg: str = next((arg for arg in argv[1:] if arg.startswith("--runs=")))
        runs: int = int(runs_arg.split("=", 1)[1])
        assert runs > 0
    except (AssertionError, StopIteration, ValueError):
        runs = 1
    elapsed: list[float] = []
    result: int | None = None
    rc: int = 0
    errors: list[str] = []
    for _ in range(runs):
        _start, _result, _stop = (perf_counter(), solve(**kwargs), perf_counter())
        elapsed.append(_stop - _start)
        if result is not None and _result != result:
            errors.append(f"Expected consistent result, got {_result} previous result={result}")
        result = _result
    if result is None:
        errors.append("Expected a result, got None")
    average: float = sum(elapsed) / len(elapsed)
    if errors:
        print("\n".join(errors), file=stderr)
        rc = 1
    print(f"{runs} {average} {result}")
    return rc


if __name__ == "__main__":
    raise SystemExit(main(multiples=int(argv[1])))
